File: src/main/java/fr/umontpellier/iut/dominion/Player/GainSkills.java
package fr.umontpellier.iut.dominion.Player;

import fr.umontpellier.iut.dominion.Destination;
import fr.umontpellier.iut.dominion.Flags;
import fr.umontpellier.iut.dominion.Item;
import fr.umontpellier.iut.dominion.Properties;
import fr.umontpellier.iut.dominion.cards.Card;
import fr.umontpellier.iut.dominion.cards.Events.Event;
import fr.umontpellier.iut.dominion.cards.Events.OnGainEvent;
import fr.umontpellier.iut.dominion.cards.component.TriggerComponent;
import javafx.beans.property.IntegerProperty;

import java.util.List;

public class GainSkills implements PlayerComponent {

    private final Player self;

    public GainSkills(Player self) {
        this.self = self;
    }

    public void gainTo(Card gainedCard, List<Card> location, Destination destination) {
        if(gainedCard == null)return;
        int i = self.getGame().tradeRoute(gainedCard);
        self.increment(Item.COIN_TOKEN_ROUTE, i);

        gainedCard.moveTo(location, destination);
    }

    public void gain(Card card, Destination dest, boolean isBuy) {
        if (card == null) return;
        Event event = new Event(card, dest, self, isBuy);

        self.checkGainToken(event);
        self.triggerOneCard(TriggerComponent.checkItselfGain.class, event.getCard(), event);
        self.triggerPlayerTavern(TriggerComponent.DuringPlayerGain.class, event);
        self.triggerEvent(TriggerComponent.DuringPlayerGain.class, event);
        self.triggerActiveEffect(TriggerComponent.DuringPlayerGain.class, event);

        if (self.getController() != self) {
            self.getController().gainTo(event.getDest(), event.getCard());
            return;
        } else {
            gainTo(event.getDest(), event.getCard());
        }

        self.getCardGainedCurrentTurn().add(event.getCard());

        if (!event.getCard().hasForLocation(event.getDest())) {
            self.moveTo(event.getCard(), event.getDest());
        }

        self.triggerEvent(TriggerComponent.AfterPlayerGain.class, event);
        self.getGame().fireEvent(OnGainEvent.class, new OnGainEvent(event));

        updateNumberOfBought(event);
    }


    public void gainSilent(Card card, Destination dest, boolean gained) {
        if(card == null)return;
        if(gained) self.getCardGainedCurrentTurn().add(card);
        if (!self.getController().equals(self)) {
            self.getController().gainTo(dest, card);
            return;
        }
        Event event = new Event(card, dest, self);
        self.checkGainToken(event);

        if(gained) self.triggerOneCard(TriggerComponent.checkItselfGain.class, card, event);
        self.getGame().fireEvent(OnGainEvent.class, new OnGainEvent(event));
        if(gained)gainTo(event.getDest(), card);

        if(event.getDest()==null)return;
        card.moveTo(self.get(dest), event.getDest());
    }

    public void gainTo(Destination dest, Card card){
        if(dest == null)return;
        gainTo(card, self.get(dest), dest);
    }

    public void updateNumberOfBought(Event event){
        if(!event.isBuy()) return;
        IntegerProperty prop = self.getProperties(Properties.Cards_Bought);
        prop.setValue(prop.getValue() + 1);
    }

}

File: src/main/java/fr/umontpellier/iut/dominion/Player/Skills/GainSkills.kt
package fr.umontpellier.iut.dominion.Player.Skills

import fr.umontpellier.iut.dominion.Annotation.AfterAllyTrigger
import fr.umontpellier.iut.dominion.Enums.Locations.Destination
import fr.umontpellier.iut.dominion.Item
import fr.umontpellier.iut.dominion.Player.Player
import fr.umontpellier.iut.dominion.Player.PlayerComponent.PlayerComponent
import fr.umontpellier.iut.dominion.Player.PlayerComponent.addCardCurrentTurn
import fr.umontpellier.iut.dominion.Properties
import fr.umontpellier.iut.dominion.cards.Card
import fr.umontpellier.iut.dominion.cards.Events.Event
import fr.umontpellier.iut.dominion.cards.Events.GainType
import fr.umontpellier.iut.dominion.cards.Events.OnGainEvent
import fr.umontpellier.iut.dominion.cards.Events.OnMoveEvent
import fr.umontpellier.iut.dominion.cards.Events.TriggerEvent
import fr.umontpellier.iut.dominion.cards.component.TriggerComponent.*
import fr.umontpellier.iut.dominion.Player.PlayerComponent.checkGainToken
import fr.umontpellier.iut.dominion.Player.PlayerComponent.controller
import kotlinx.coroutines.flow.MutableStateFlow

class GainSkills : PlayerComponent {
}

private fun Player.gainTo(gainedCard : Card?, location: MutableStateFlow<List<Card>>?, destination: Destination?) {
    if(gainedCard == null) return
    val i = game.tradeRoute(gainedCard)
    increment(Item.COIN_TOKEN_ROUTE, i)

    gainedCard.moveTo(location, destination)
}


suspend fun Player.gain(event: Event){ gain(event.card, event.destination, event.isBuy) }


@AfterAllyTrigger
@JvmOverloads
suspend fun Player.gain(card: Card?, dest: Destination? = Destination.PlayerZone.Discard, isBuy: Boolean = false) {
    if (card == null) return
    val event = Event(card, dest, self, isBuy)
    val onMoveEvent = OnMoveEvent(event, GainType.BEFORE)

    fun updateNumberOfBought() {
        if (!event.isBuy) return
        val prop = getProperties(Properties.Cards_Bought)
        prop.value += 1
    }

    if (controller !== self) {
        controller.gainTo(onMoveEvent.destination, onMoveEvent.card)
        return
    } else {
        gainTo(onMoveEvent.destination, onMoveEvent.card)
    }

    checkGainToken(event)
    game.fireEvent<OnMoveEvent>(onMoveEvent)

    triggerOneCard<CheckItselfGain>( event.card, event)

    event.updateGainType(GainType.DURING)

    val triggerEvent = TriggerEvent(onMoveEvent)

    triggerPlayerTavern<DuringPlayerGain>(triggerEvent)
    triggerEvent<DuringPlayerGain>(triggerEvent)
    triggerActiveEffect<SideEffectGain>(triggerEvent)

    triggerEvent.updateGainType(GainType.AFTER)

    addCardCurrentTurn(event.card)

    triggerEvent.card?.hasForLocation(triggerEvent.destination)?.let {
        if (!it) {
            moveTo(triggerEvent.card, triggerEvent.destination)
        }
    }

    triggerEvent<AfterPlayerGain>(triggerEvent)
    triggerPlayerTavern<AfterPlayerGain>(triggerEvent)
    triggerActiveEffect<SideEffectGain>(triggerEvent)
    game.fireEvent<OnGainEvent>(OnGainEvent(triggerEvent))

    updateNumberOfBought()
}


suspend fun Player.gainSilent(card: Card?, dest: Destination, gained: Boolean){
    if (card == null) return
    if(gained) addCardCurrentTurn(card)

    val event = Event(card, dest, self)

    if(controller != self){
        controller.gainTo(dest, card)
        return
    }else if (gained) gainTo(dest, event.card)

    checkGainToken(event)

    if(gained) triggerOneCard<CheckItselfGain>( card, event)
    game.fireEvent<OnGainEvent>(OnGainEvent(event))


    event.destination?.let { destination ->
        moveTo(card, destination)
    }
}

private fun Player.gainTo(dest : Destination?, c : Card?) {
    if(c == null) return

    gainTo(c, get(dest), dest)
}